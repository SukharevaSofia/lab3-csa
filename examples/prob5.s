jump _MAIN

; notation
; input:
; r0 - number1
; r1 - number2
; output:
; r0 - result
; clobbers:
; r0 - r2
_NOD:
  jeq r0 _NOD_END
  jeq r1 _NOD_END
  mov r2 r0
  sub r2 r1
  jge r2 _NOD_POS
  jump _NOD_NEG
_NOD_POS:
  mod r0 r1
  jump _NOD
_NOD_NEG:
  mod r1 r0
  jump _NOD
_NOD_END:
  mov r2 r0
  sub r2 r1
  jge r2 _NOD_RET
  mov r0 r1
_NOD_RET:
  jump _NOK_FROM_NOD


; inputs:
; r3 - number1
; r4 - number2
; return:
; r3 res
; clobbers:
; r0-r5
_NOK:
mov r0 r3
mov r1 r4
jump _NOD
_NOK_FROM_NOD:
mul r3 r4
div r3 r0
jump _MAIN_LOOP_RET

; r7 res
; r8 iter
_MAIN:
ldc r7 1
ldc r8 1
_MAIN_LOOP:
inc r8
mov r3 r7
mov r4 r8
jump _NOK
_MAIN_LOOP_RET:
mov r7 r3
ldc r9 20
sub r9 r8
jge r9 _MAIN_LOOP

outn r7
ldc r7 10
out r7
hlt
