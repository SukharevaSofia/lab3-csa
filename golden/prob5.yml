source: |-
  jump _MAIN

  ; notation
  ; input:
  ; r0 - number1
  ; r1 - number2
  ; output:
  ; r0 - result
  ; clobbers:
  ; r0 - r2
  _NOD:
    jeq r0 _NOD_END
    jeq r1 _NOD_END
    mov r2 r0
    sub r2 r1
    jge r2 _NOD_POS
    jump _NOD_NEG
  _NOD_POS:
    mod r0 r1
    jump _NOD
  _NOD_NEG:
    mod r1 r0
    jump _NOD
  _NOD_END:
    mov r2 r0
    sub r2 r1
    jge r2 _NOD_RET
    mov r0 r1
  _NOD_RET:
    jump _NOK_FROM_NOD


  ; inputs:
  ; r3 - number1
  ; r4 - number2
  ; return:
  ; r3 res
  ; clobbers:
  ; r0-r5
  _NOK:
  mov r0 r3
  mov r1 r4
  jump _NOD
  _NOK_FROM_NOD:
  mul r3 r4
  div r3 r0
  jump _MAIN_LOOP_RET

  ; r7 res
  ; r8 iter
  _MAIN:
  ldc r7 1
  ldc r8 1
  _MAIN_LOOP:
  inc r8
  mov r3 r7
  mov r4 r8
  jump _NOK
  _MAIN_LOOP_RET:
  mov r7 r3
  ldc r9 20
  sub r9 r8
  jge r9 _MAIN_LOOP

  outn r7
  ldc r7 10
  out r7
  hlt



stdin: '123123123123'

target: |-
  {"program": [{"instr": "jump", "args": [22]}, {"instr": "jeq", "args": [0, 11]}, {"instr": "jeq", "args": [1, 11]}, {"instr": "mov", "args": [2, 0]}, {"instr": "sub", "args": [2, 1]}, {"instr": "jge", "args": [2, 7]}, {"instr": "jump", "args": [9]}, {"instr": "mod", "args": [0, 1]}, {"instr": "jump", "args": [1]}, {"instr": "mod", "args": [1, 0]}, {"instr": "jump", "args": [1]}, {"instr": "mov", "args": [2, 0]}, {"instr": "sub", "args": [2, 1]}, {"instr": "jge", "args": [2, 15]}, {"instr": "mov", "args": [0, 1]}, {"instr": "jump", "args": [19]}, {"instr": "mov", "args": [0, 3]}, {"instr": "mov", "args": [1, 4]}, {"instr": "jump", "args": [1]}, {"instr": "mul", "args": [3, 4]}, {"instr": "div", "args": [3, 0]}, {"instr": "jump", "args": [28]}, {"instr": "ldc", "args": [7, 1]}, {"instr": "ldc", "args": [8, 1]}, {"instr": "inc", "args": [8]}, {"instr": "mov", "args": [3, 7]}, {"instr": "mov", "args": [4, 8]}, {"instr": "jump", "args": [16]}, {"instr": "mov", "args": [7, 3]}, {"instr": "ldc", "args": [9, 20]}, {"instr": "sub", "args": [9, 8]}, {"instr": "jge", "args": [9, 24]}, {"instr": "outn", "args": [7]}, {"instr": "ldc", "args": [7, 10]}, {"instr": "out", "args": [7]}, {"instr": "hlt", "args": []}], "data": []}
stdout: "232792560\n"

logs: ./golden/logs/prob5.log
